### PR TITLE
feat: thorchain lp sweep step

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

This brings the sweep step for UTXOs, as well as required gas estimation / balance for gas checks groundwork (at least for UTXOs as a pool asset, i.e not for ROON yet, and not *proper* gas estimation of `depositWithExpiry` calldata for EVMs just yet yet), which we'll be able to iterate on in a follow-up and do full-featured gas estimations and surface them in the UI in a follow-up.

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

- contributes the sweep step of https://github.com/shapeshift/web/issues/5960, and fee estimation/checks on the asset side, the logic of which will be able to be reused in a follow-up PR to surface fees checks in the UI (i.e missing balance for gas) as well as display fees 

- This brings regressions risks to the approval feature right under this PR. Ensure it's still happy, as well as balance checks.

> What protocols, transaction types or contract interactions might be affected by this PR?

None, this sends fundus to yourselfu

## Testing

- Send your whole balance to your next receive address (or if you want to be fancy, ensure your xpub's 0th address has less than needed for your deposit)
- Ensure sweep step is detected as needed, is triggered, and sweep is happy
- Ensure you can proceed with the deposit

- With enough balance in your xpub's 0th address for depositing, ensure sweep step isn't triggered
- Ensure other chains aren't affected by any regression from this
- Ensure approvals / allowance checks are still happy for EVMs

### Engineering

- ☝🏽 

### Operations

- ☝🏽 

## Screenshots (if applicable)

- Sweep

https://github.com/shapeshift/web/assets/17035424/b4e50c91-cd6a-479e-aca3-f990ff4168f3

- Approvals are still happy

https://github.com/shapeshift/web/assets/17035424/3b0daa66-9223-4008-b8a5-7927182ba246

### CODE REVIEW

**Comment by gomesalexandre** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/shapeshift/web/6110?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `develop`
  * **PR #6106** <a href="https://app.graphite.dev/github/pr/shapeshift/web/6106?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #6110** <a href="https://app.graphite.dev/github/pr/shapeshift/web/6110?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/6110?utm_source=stack-comment).

File: src/pages/ThorChainLP/components/AddLiquitity/AddLiquidityInput.tsx [hunk 1]
```diff
@@ -311,61 +321,59 @@ export const AddLiquidityInput: React.FC<AddLiquidityInputProps> = ({
     [asset, accountIdsByChainId],
   )
 
-  const assetBalanceCryptoBaseUnit = useAppSelector(state =>
+  const poolAssetBalanceCryptoBaseUnit = useAppSelector(state =>
     selectPortfolioCryptoBalanceBaseUnitByFilter(state, assetBalanceFilter),
   )
+  const poolAssetFeeAsset = useAppSelector(state => selectFeeAssetById(state, asset?.assetId ?? ''))
```
Comment (gomesalexandre) [inline src/pages/ThorChainLP/components/AddLiquitity/AddLiquidityInput.tsx:327]: Note `poolAsset` terminology - we will still need to do RUNE gas estimations, with the exact same logic on the RUNE side but that's not related to this PR

File: src/pages/ThorChainLP/components/AddLiquitity/AddLiquidityInput.tsx [hunk 2]
```diff
@@ -375,7 +383,7 @@ export const AddLiquidityInput: React.FC<AddLiquidityInputProps> = ({
     ...reactQueries.mutations.approve({
       assetId: asset?.assetId,
       spender: inboundAddressData?.router,
-      from: allowanceFromAddress,
+      from: poolAssetAccountAddress,
```
Comment (gomesalexandre) [inline src/pages/ThorChainLP/components/AddLiquitity/AddLiquidityInput.tsx:386]: Unified into `poolAssetAccountAddress` - since we do need  a from address for all pool assets, not only for EVM tokens in the context of approvals, we don't need `allowanceFromAddress` anymore as that's effectively duplicate logic

File: src/pages/ThorChainLP/components/AddLiquitity/AddLiquidityInput.tsx [hunk 3]
```diff
@@ -435,15 +443,150 @@ export const AddLiquidityInput: React.FC<AddLiquidityInputProps> = ({
...
+
+    if (!transactionType) return
+
+    switch (transactionType) {
+      case 'MsgDeposit': {
+        return THORCHAIN_POOL_MODULE_ADDRESS
+      }
+      case 'EvmCustomTx': {
+        // TODO: this should really be inboundAddressData?.router, but useQuoteEstimatedFeesQuery doesn't yet handle contract calls
+        // for the purpose of naively assuming a send, using the inbound address instead of the router is fine
```
Comment (gomesalexandre) [inline src/pages/ThorChainLP/components/AddLiquitity/AddLiquidityInput.tsx:493]: Pretty much what this says - not for this PR

File: src/react-queries/hooks/useQuoteEstimatedFeesQuery.ts [hunk 4]
```diff
@@ -27,6 +39,14 @@ type UseQuoteEstimatedFeesProps = {
       repaymentAccountId: AccountId
       repaymentAsset: Asset | null
     }
+  | {
+      confirmedQuote: LpConfirmedDepositQuote | null
+      // Technically not a collateral, but this avoids weird branching, ternaries or ?? check sfor now
```
Comment (gomesalexandre) [inline src/react-queries/hooks/useQuoteEstimatedFeesQuery.ts:44]: If we really don't like that smell of this, this can be cleaned up and make `const accountId = repaymentAccountId ?? collateralAccountId` an IIFE instead

File: src/react-queries/hooks/useQuoteEstimatedFeesQuery.ts [hunk 5]
```diff
@@ -47,11 +67,16 @@ export const useQuoteEstimatedFeesQuery = ({
     const supportedEvmChainIds = getSupportedEvmChainIds()
     const cryptoAmount = depositAmountCryptoPrecision ?? repaymentAmountCryptoPrecision ?? '0'
     const assetId = repaymentAsset?.assetId ?? collateralAssetId
-    const quoteMemo = confirmedQuote?.quoteMemo ?? confirmedQuote?.quoteMemo ?? ''
```
Comment (gomesalexandre) [inline src/react-queries/hooks/useQuoteEstimatedFeesQuery.ts:50]: wat was this
Added via Giphy

File: src/react-queries/hooks/useQuoteEstimatedFeesQuery.ts [hunk 6]
```diff
@@ -47,11 +67,16 @@ export const useQuoteEstimatedFeesQuery = ({
...
-    const memo = supportedEvmChainIds.includes(fromAssetId(assetId).chainId as KnownChainIds)
-      ? utils.hexlify(utils.toUtf8Bytes(quoteMemo))
-      : quoteMemo
-    const to = confirmedQuote?.quoteInboundAddress ?? confirmedQuote?.quoteInboundAddress ?? ''
```
Comment (gomesalexandre) [inline src/react-queries/hooks/useQuoteEstimatedFeesQuery.ts:54]: and wat was that
Added via Giphy

File: src/react-queries/hooks/useQuoteEstimatedFeesQuery.ts (hunk 5 above)
Comment (0xApotheosis) [reply src/react-queries/hooks/useQuoteEstimatedFeesQuery.ts:50]: lol, you tell me, good ser! 😂

File: src/pages/ThorChainLP/components/AddLiquitity/AddLiquidityInput.tsx (hunk 2 above)
Comment (0xApotheosis) [reply src/pages/ThorChainLP/components/AddLiquitity/AddLiquidityInput.tsx:386]: 🧠

**Review by 0xApotheosis — APPROVED** [review]
✅ Successful Litecoin consolidation, which was correctly triggered when necessary: https://live.blockcypher.com/ltc/tx/06f45211d88141b5224dbc3420486dc965f7ba99e4345f72b286e4603bf230d5/

✅ Confirming consolidation was not triggered when not necessary.

🤔 
Screenshot 2024-02-05 at 11 01 48 am

Even once the consolidation TX was complete, and had 6+ confirmations the UI did not update accordingly. I suspect because https://dev-daemon.thorchain.shapeshift.com/lcd/thorchain/tx/status/06f45211d88141b5224dbc3420486dc965f7ba99e4345f72b286e4603bf230d5 is still showing as `completed: false` at time of testing.

File: src/pages/ThorChainLP/components/AddLiquitity/AddLiquidityInput.tsx [hunk 7]
```diff
@@ -435,15 +443,150 @@ export const AddLiquidityInput: React.FC<AddLiquidityInputProps> = ({
...
+  const isSweepNeededArgs = useMemo(
+    () => ({
+      assetId: asset?.assetId,
+      address: poolAssetAccountAddress ?? null,
+      amountCryptoBaseUnit: toBaseUnit(
+        actualAssetCryptoLiquidityAmount ?? 0,
+        asset?.precision ?? 0,
+      ),
+      // Effectively defined at runtime because of the enabled check below
+      txFeeCryptoBaseUnit: estimatedFeesData?.txFeeCryptoBaseUnit!,
```
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/pages/ThorChainLP/components/AddLiquitity/AddLiquidityInput.tsx:556]: Only a blocker if I've understood that correctly.

File: src/react-queries/hooks/useQuoteEstimatedFeesQuery.ts (hunk 5 above)
Comment (gomesalexandre) [reply src/react-queries/hooks/useQuoteEstimatedFeesQuery.ts:50]: > lol, you tell me, good ser! 😂

don't try blaming those lines, this is showing me as the author which must be an obvious bug in git core

File: src/pages/ThorChainLP/components/AddLiquitity/AddLiquidityInput.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/pages/ThorChainLP/components/AddLiquitity/AddLiquidityInput.tsx:556]: image
`txFeeCryptoBaseUnit` should be defined so long as we have `estimatedFeesData` but doesn't hurt to err on the side of paranoia 🙏🏽  - fa2404bc8fc81e3eb5f57b6d8a70cb9155ca0154

File: src/pages/ThorChainLP/components/AddLiquitity/AddLiquidityInput.tsx [hunk 8]
```diff
@@ -435,15 +443,150 @@ export const AddLiquidityInput: React.FC<AddLiquidityInputProps> = ({
...
+        wallet,
+        accountMetadata: poolAccountMetadata,
+        getPosition: getThorchainLpPosition,
+      })
+      setPoolAssetAccountAddress(_accountAssetAddress)
+    })()
+  }, [activeOpportunityId, asset, poolAccountId, poolAccountMetadata, wallet])
+
+  const poolAssetInboundAddress = useMemo(() => {
+    const transactionType = (() => {
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/pages/ThorChainLP/components/AddLiquitity/AddLiquidityInput.tsx]: Absolutely - 1206fac8afba165717abaa453227cac561f4192f

**Comment by gomesalexandre** [comment]
> 🤔 Even once the consolidation TX was complete, and had 6+ confirmations the UI did not update accordingly. I suspect because https://dev-daemon.thorchain.shapeshift.com/lcd/thorchain/tx/status/06f45211d88141b5224dbc3420486dc965f7ba99e4345f72b286e4603bf230d5 is still showing as `completed: false` at time of testing.

tl;dr Litecoin currently slow, need more awaiting before fetching UTXOs

Consolidation isn't related to THOR so we wouldn't be able to see the Tx in THOR endpoints - the only thing sweep does is it send all your funds to a specific address (most likely your 0th address as active or default address, although if you deposited with our old savers implementation, your active addres could well be an address at 0+ position), and wait for the Txid to be seen in your list of UTXOs *for that specfic address, not for the xpub* after a 15s timeout to ensure things are propagated across our different load-balanced nodes.

I've tested LTC consolidation locally myself, and indeed, it spinned infinitely, as the call to the /utxos endpoint was returning an empty array. Manually refetching the request a few seconds later made it happy, so I've bumped the sleep time to 60 instead of 15 to handle chains where the Txid takes longer to propagate 🙏🏽. Tried 30 seconds initially, but that didn't seem to cut it  so 60 it is - ae02f584ce1c4eacc8ea492561626ee331198f79

https://github.com/shapeshift/web/assets/17035424/a77968db-7dfa-424e-b4bd-ef3c3a32874e
